tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+
+		// roll over a new segment file so we can compact all the old ones
+		if err := p.newSegmentFile(); err != nil {
+			return err
+		}
+		compactFilesLessThan = p.currentSegmentID
+
+		return nil
+	}(); err == ErrCompactionRunning {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// write the data to the index first
+	if err := idx.WriteIndex(seriesToFlush); err != nil {
+		// if we can't write the index, we should just bring down the server hard
+		panic(fmt.Sprintf("error writing the wal to the index: %s", err.Error()))
+	}
+
+	// clear the flush cache
+	p.mu.Lock()
+	p.flushCache = nil
+	p.mu.Unlock()
+
+	// now compact all the old data
+	fileNames, err := p.segmentFileNames()
+	if err != nil {
+		return err
+	}
+
+	// all compacted data from the segments will go into this file
+	compactionFile, err := os.OpenFile(p.compactionFileName(), os.O_CREATE|os.O_RDWR, 0600)
+	if err != nil {
+		return err
+	}
+
+	for _, n := range fileNames {
+		id, err := p.idFromFileName(n)
+		if err != nil {
+			return err
+		}
+
+		// only compact files that are older than the segment that became active when we started the flush
+		if id >= compactFilesLessThan {
+			break
+		}
+
+		f, err := os.OpenFile(n, os.O_RDONLY, 0600)
+		if err != nil {
+			return err
+		}
+
+		sf := newSegmentFile(f)
+		var entries []*entry
+		for {
+			a, err := sf.readCompressedBlock()
+
+			if err == ErrCompactionBlock {
+				continue
+			} else if err != nil {
+				return err
+			}
+
+			if a == nil {
+				break
+			}
+
+			// only compact the entries from series that haven't been flushed
+			for _, e := range a {
+				if _, ok := seriesToFlush[string(e.key)]; !ok {
+					entries = append(entries, e)
+				}
+			}
+		}
+
+		if err := p.writeCompactionEntry(compactionFile, entries); err != nil {
+			return err
+		}
+
+		// now close and delete the file
+		if err := f.Close(); err != nil {
+			return err
+		}
+		os.Remove(n)
+	}
+
+	if err := compactionFile.Close(); err != nil {
+		return err
+	}
+
+	// close the compaction file and rename it so that it will appear as the very first segment
+	compactionFile.Close()
+	os.Rename(compactionFile.Name(), p.fileNameForSegment(1))
+
+	// and mark the compaction as done
+	p.mu.Lock()
+	p.compactionRunning = false
+	p.mu.Unlock()
+
+	return nil
+}
+
+// writeCompactionEntry will write a marker for the beginning of the file we're compacting, a compressed block
+// for all entries, then a marker for the end of the file
+func (p *partition) writeCompactionEntry(f *os.File, entries []*entry) error {
+	if err := p.writeCompactionFileName(f); err != nil {
+		return err
+	}
+
+	block := make([]byte, 0)
+	for _, e := range entries {
+		block = append(block, marshalWALEntry(e.key, e.timestamp, e.data)...)
+	}
+	b := snappy.Encode(nil, block)
+
+	if n, err := f.Write(u64tob(uint64(len(b)))); err != nil {
+		return err
+	} else if n != 8 {
+		return fmt.Errorf("compaction expected to write %d bytes but wrote %d", 8, n)
+	}
```

Comment:
You can use [`binary.Write()`](https://golang.org/pkg/encoding/binary/#Write) which will ensure the appropriate length is written so you can just check the error.